### PR TITLE
BUG: iloc.__setitem__ with duplicate columns

### LINKED ISSUE #15686: BUG: .iloc indexing with duplicates
#### Code Sample, a copy-pastable example if possible

```python
In [2]: df1 = pd.DataFrame([{'A': None, 'B': 1}, {'A': 2, 'B': 2}])
   ...: df2 = pd.DataFrame([{'A': 3, 'B': 3}, {'A': 4, 'B': 4}])
   ...: df = pd.concat([df1, df2], axis=1)
   ...: df.iloc[0, 0] = df.iloc[0, 2]
   ...: df
   ...: 
Out[2]: 
     A  B    A  B
0  3.0  1  3.0  3
1  2.0  2  4.0  4
```
#### Problem description

The third column shouldn't have changed type

#### Expected Output
``` python
Out[2]: 
     A  B  A  B
0  3.0  1  3  3
1  2.0  2  4  4
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: 32df1e6ae452f7ddd31dc41fa613992493eb51c4
python: 3.5.3.final.0
python-bits: 64
OS: Linux
OS-release: 4.7.0-1-amd64
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: it_IT.utf8
LOCALE: it_IT.UTF-8

pandas: 0.19.0+597.g32df1e6ae
pytest: 3.0.6
pip: 9.0.1
setuptools: 33.1.1
Cython: 0.25.2
numpy: 1.12.0
scipy: 0.18.1
xarray: None
IPython: 5.1.0.dev
sphinx: 1.4.9
patsy: 0.3.0-dev
dateutil: 2.5.3
pytz: 2016.7
blosc: None
bottleneck: 1.2.0
tables: 3.3.0
numexpr: 2.6.1
feather: 0.3.1
matplotlib: 2.0.0
openpyxl: 2.3.0
xlrd: 1.0.0
xlwt: 1.1.2
xlsxwriter: 0.9.6
lxml: 3.7.1
bs4: 4.5.3
html5lib: 0.999999999
sqlalchemy: 1.0.15
pymysql: None
psycopg2: None
jinja2: 2.8
s3fs: None
pandas_gbq: None
pandas_datareader: 0.2.1

</details>

### LINKED ISSUE #22036: Replacing a column with iloc replaces another column with same name
#### Code Sample, a copy-pastable example if possible

```python
import pandas

a = pandas.DataFrame({'a': ['0'], 'b': ['str']})

print('---')
print(a)

a.iloc[:, 0] = [int(v) for v in a.iloc[:, 0]]

print('---')
print(a)

b = pandas.concat([a, pandas.DataFrame({'b': ['str2']})], axis=1)

print('---')
print(b)

b.iloc[:, 2] = ['str3']

print('---')
print(b)
```

#### Problem description

The issue seems to be that if there is a DataFrame with duplicate column names and mixed dtypes, if I try to replace one column with another value, using `iloc`, another column with same name is replaced as well.

#### Expected Output

The final print should be:

```
   a    b     b
0  0  str  str3
```

And not:

```
   a     b     b
0  0  str3  str3
```

It is interesting that if I change concat line to (see renaming of column `b` to `c`):

```
b = pandas.concat([a, pandas.DataFrame({'c': ['str2']})], axis=1)
```

Then the output is correctly:

```
   a    b     c
0  0  str  str3
```

Also, if `a.iloc[:, 0] = [int(v) for v in a.iloc[:, 0]]` is commented out, it works also correctly.

Moreover, the following also work correctly (see the change of `0` column index into `[0]` column index, and similar for `2` (this latter change is not really necessary to make it work)):

```python
import pandas

a = pandas.DataFrame({'a': ['0'], 'b': ['str']})

print('---')
print(a)

a.iloc[:, [0]] = [int(v) for v in a.iloc[:, 0]]

print('---')
print(a)

b = pandas.concat([a, pandas.DataFrame({'b': ['str2']})], axis=1)

print('---')
print(b)

b.iloc[:, [2]] = ['str3']

print('---')
print(b)
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.3.final.0
python-bits: 64
OS: Linux
OS-release: 4.13.0-46-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.23.3
pytest: None
pip: 18.0
setuptools: 40.0.0
Cython: None
numpy: 1.15.0
scipy: None
pyarrow: None
xarray: None
IPython: None
sphinx: None
patsy: None
dateutil: 2.7.3
pytz: 2018.5
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: None
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #15686
- [x] closes #22036
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/indexing.py [hunk 1]
```diff
@@ -1615,6 +1615,12 @@ def _setitem_with_indexer(self, indexer, value):
                 info_idx = [info_idx]
             labels = item_labels[info_idx]
 
+            # Ensure we have something we can iterate over
+            ilocs = info_idx
+            if isinstance(info_idx, slice):
+                ri = Index(range(len(self.obj.columns)))
```
Comment (jreback) [inline pandas/core/indexing.py]: use np.arange

File: pandas/core/indexing.py [hunk 2]
```diff
@@ -1632,9 +1638,10 @@ def _setitem_with_indexer(self, indexer, value):
                         "length than the value"
                     )
 
-            def setter(item, v):
-                ser = self.obj[item]
-                pi = plane_indexer[0] if lplane_indexer == 1 else plane_indexer
+            pi = plane_indexer[0] if lplane_indexer == 1 else plane_indexer
+
+            def isetter(loc, v):
+                ser = self.obj._ixs(loc, axis=1)
```
Comment (jreback) [inline pandas/core/indexing.py]: can you comment that this is not a positional setter

File: pandas/core/frame.py [hunk 3]
```diff
@@ -2706,6 +2706,20 @@ def _setitem_frame(self, key, value):
         self._check_setitem_copy()
         self._where(-key, value, inplace=True)
 
+    def _iset_item(self, loc: int, value):
+        self._ensure_valid_index(value)
+
+        # passing loc to sanitize_column is a misnomer, but harmless
+        #  with broadcast=False as long as value is never a DataFrame
```
Comment (jreback) [inline pandas/core/frame.py]: not sure i unserstand this comment

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexing.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/indexing.py]: np.arange will give us an Int64Index, whereas this will give us a RangeIndex

**Comment by jbrockmendel** [comment]
updated+green

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks!

**Pull request merged**